### PR TITLE
Support run as windows service

### PR DESCRIPTION
**Pull request by chai2010**
1. `sc create etcd binPath= "C:\etcd\etcd.exe -data-dir=C:\etcd\etcd.service.data"`
2. `net start etcd`
3. `net stop etcd`
4. `sc delete etcd`

Notes: Add "golang.org/x/sys/windows/svc" to vendor.

### CODE REVIEW

**Comment by xiang90** [comment]
@chai2010 Can you fix the issue reported by go vet? `main_windows.go:143: declaration of err shadows declaration at main_windows.go:130: `

@judwhite @cloudtrends Can you please take a look?

**Comment by judwhite** [comment]
I'm very interested in this, just wanted to give a heads up my response may be delayed due to other commitments. I will try it out locally and follow up within the week.

**Comment by judwhite** [comment]
I have some concerns with the approach taken here. I'll leave comments on this PR but it may require enough changes where a new PR might make more sense. Thanks @chai2010 for getting this started.

File: main_windows.go [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+	}
+	etcdmain.Main()
+}
+
+func (m *etcdService) StopServer() {
+	//
+}
+
+func (m *etcdService) Execute(args []string, r <-chan svc.ChangeRequest, changes chan<- svc.Status) (ssec bool, errno uint32) {
+	const cmdsAccepted = svc.AcceptStop | svc.AcceptShutdown | svc.AcceptPauseAndContinue
```
Comment (judwhite) [inline main_windows.go]: Suggest removing `svc.AcceptPauseAndContinue`

File: main_windows.go [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+				changes <- c.CurrentStatus
+				// testing deadlock from https://code.google.com/p/winsvc/issues/detail?id=4
+				time.Sleep(100 * time.Millisecond)
+				changes <- c.CurrentStatus
+			case svc.Stop, svc.Shutdown:
+				break loop
+			case svc.Pause:
+				changes <- svc.Status{State: svc.Paused, Accepts: cmdsAccepted}
+			case svc.Continue:
+				changes <- svc.Status{State: svc.Running, Accepts: cmdsAccepted}
```
Comment (judwhite) [inline main_windows.go]: Suggest removing `Pause` and `Continue`

File: main_windows.go [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+			default:
+				m.elog.Error(1, fmt.Sprintf("etcd: unexpected control request #%d", c))
```
Comment (judwhite) [inline main_windows.go]: `continue loop` should be sufficient

File: main_windows.go [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+loop:
+	for {
+		select {
+		case c := <-r:
+			switch c.Cmd {
+			case svc.Interrogate:
+				changes <- c.CurrentStatus
+				// testing deadlock from https://code.google.com/p/winsvc/issues/detail?id=4
+				time.Sleep(100 * time.Millisecond)
+				changes <- c.CurrentStatus
```
Comment (judwhite) [inline main_windows.go]: Take a look at the issue listed, I believe this was added to their example just for testing that issue. I don't think the comment and the two lines after are necessary.

File: main_windows.go [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+	}
+	if err := os.Chdir(filepath.Dir(exepath)); err != nil {
+		log.Fatal(err)
+	}
+	etcdmain.Main()
+}
+
+func (m *etcdService) StopServer() {
+	//
+}
```
Comment (judwhite) [inline main_windows.go]: We should try to do a clean shutdown. It looks like `RegisterInterruptHandler` is called to add a graceful shutdown process, but it's currently a no-op on Windows. I think `interrupt_windows.go` could be modified to look more like `interrupt_unix.go`.

I just tried `Ctrl+C` on Windows and it looked to stop the process hard.

**Comment by judwhite** [comment]
@chai2010 Are you still interested in this? If you're too busy I can see it through and find another reviewer to help us out.

File: main_windows.go (hunk 1 above)
Comment (chai2010) [inline main_windows.go]: Done

File: main_windows.go (hunk 2 above)
Comment (chai2010) [inline main_windows.go]: Done

File: main_windows.go (hunk 3 above)
Comment (chai2010) [inline main_windows.go]: The origin code is just copy from the winsvc example, i don't understand the every line clearly.
I agree with you.

File: main_windows.go (hunk 5 above)
Comment (chai2010) [inline main_windows.go]: If the `etcd` is not in `windows service` mode, the `CTRL+C` is OK (just use `signal.Notify`).
If the `etcd` is run as `windows service`, i don't know how to work with `CTRL+C`.
Need help.

**Comment by chai2010** [comment]
@judwhite sorry for this delayed comments. I do some other work these days. 
Thanks for your review.

**Comment by hminaya** [comment]
Any updates on this?

**Comment by judwhite** [comment]
Sort of, I'm working with the authors of NSQ on a similar change. We're working out final details and I'd like to reference the work there once it's merged. The PR is old, but day jobs, and holidays. We're picking it back up and just spoke today about it. https://github.com/nsqio/nsq/pull/676

**Comment by hminaya** [comment]
@judwhite great, thanks for the update, if there is anything I can do to help just let me know

**Comment by judwhite** [comment]
This landed in https://github.com/nsqio/nsq/pull/718

@chai2010 @xiang90 @hminaya what do you think of this approach?

**Comment by hminaya** [comment]
@judwhite shounds good to me.

**Comment by gyuho** [comment]
Closing due to inactivity. Please create a new one if needed.

**Pull request closed without merging**